File: app/src/main/java/com/miaadrajabi/menuxmlparser/java/data/repository/MenuRepository.java
package com.miaadrajabi.menuxmlparser.java.data.repository;

import com.miaadrajabi.menuxmlparser.java.data.mapper.XmlMappers;
import com.miaadrajabi.menuxmlparser.java.data.source.AssetsXmlDataSource;
import com.miaadrajabi.menuxmlparser.java.data.source.LocalXmlDataSource;
import com.miaadrajabi.menuxmlparser.java.data.xml.XmlModels;
import com.miaadrajabi.menuxmlparser.java.domain.model.MenuModels.DynamicMenus;

import java.io.IOException;

public class MenuRepository {
    private final AssetsXmlDataSource assets;
    private final LocalXmlDataSource local;

    public MenuRepository(AssetsXmlDataSource assets, LocalXmlDataSource local) {
        this.assets = assets;
        this.local = local;
    }

    public DynamicMenus load() {
        try {
            XmlModels fromLocal = local.read();
            XmlModels xml = fromLocal != null ? fromLocal : assets.read();
            return XmlMappers.toDomain(xml);
        } catch (IOException e) {
            throw new RuntimeException("Failed to load menu data", e);
        }
    }

    public void save(DynamicMenus domain) {
        try {
            XmlModels xml = toXml(domain);
            local.write(xml);
        } catch (IOException e) {
            throw new RuntimeException("Failed to save menu data", e);
        }
    }

    // Reverse mappers for persistence
    private XmlModels toXml(DynamicMenus domain) {
        XmlModels xml = new XmlModels();
        for (com.miaadrajabi.menuxmlparser.java.domain.model.MenuModels.MenuGroup g : domain.groups) {
            xml.menuGroups.add(toXml(g));
        }
        return xml;
    }

    private XmlModels.MenuGroupXml toXml(com.miaadrajabi.menuxmlparser.java.domain.model.MenuModels.MenuGroup g) {
        XmlModels.MenuGroupXml gx = new XmlModels.MenuGroupXml();
        gx.clazz = null;
        gx.enTitle = g.englishTitle;
        gx.enable = Boolean.toString(g.isEnabled);
        gx.faTitle = g.persianTitle;
        gx.id = g.id;
        for (com.miaadrajabi.menuxmlparser.java.domain.model.MenuModels.TopLevelMenu t : g.topLevelMenus) {
            gx.topLevelMenus.add(toXml(t));
        }
        return gx;
    }

    private XmlModels.TopLevelMenuXml toXml(com.miaadrajabi.menuxmlparser.java.domain.model.MenuModels.TopLevelMenu t) {
        XmlModels.TopLevelMenuXml tx = new XmlModels.TopLevelMenuXml();
        tx.enTitle = t.englishTitle;
        tx.enable = Boolean.toString(t.isEnabled);
        tx.expandable = t.isExpandable != null ? Boolean.toString(t.isExpandable) : null;
        tx.faTitle = t.persianTitle;
        tx.icon = t.icon;
        tx.id = t.id;
        for (com.miaadrajabi.menuxmlparser.java.domain.model.MenuModels.ServiceDescriptor s : t.services) {
            tx.services.add(toXml(s));
        }
        for (com.miaadrajabi.menuxmlparser.java.domain.model.MenuModels.MenuItem m : t.items) {
            tx.items.add(toXml(m));
        }
        return tx;
    }

    private XmlModels.MenuItemXml toXml(com.miaadrajabi.menuxmlparser.java.domain.model.MenuModels.MenuItem m) {
        XmlModels.MenuItemXml mx = new XmlModels.MenuItemXml();
        mx.clazz = m.clazz;
        mx.enTitle = m.englishTitle;
        mx.enable = Boolean.toString(m.isEnabled);
        mx.faTitle = m.persianTitle;
        mx.id = m.id;
        mx.ussdGetSimCharge = m.ussdGetSimCharge;
        mx.ussdGetSimNum = m.ussdGetSimNum;
        mx.usssdChargeCommand = m.usssdChargeCommand;
        for (com.miaadrajabi.menuxmlparser.java.domain.model.MenuModels.MenuItem c : m.children) {
            mx.children.add(toXml(c));
        }
        for (com.miaadrajabi.menuxmlparser.java.domain.model.MenuModels.ServiceDescriptor s : m.services) {
            mx.services.add(toXml(s));
        }
        return mx;
    }

    private XmlModels.ServiceDescriptorXml toXml(com.miaadrajabi.menuxmlparser.java.domain.model.MenuModels.ServiceDescriptor s) {
        XmlModels.ServiceDescriptorXml sx = new XmlModels.ServiceDescriptorXml();
        sx.id = s.id;
        sx.service = s.service;
        sx.serviceAmount = s.serviceAmount;
        sx.categoryId = s.categoryId;
        sx.hasCount = s.hasCount;
        sx.load = s.load;
        sx.providerId = s.providerId;
        sx.support = s.support;
        sx.algorithmType = s.algorithmType;
        sx.companyNameEn = s.companyNameEn;
        sx.companyNameFa = s.companyNameFa;
        return sx;
    }
}

File: app/src/main/java/com/miaadrajabi/menuxmlparser/data/repository/MenuRepository.kt
package com.miaadrajabi.menuxmlparser.data.repository

import com.miaadrajabi.menuxmlparser.data.mapper.toDomain
import com.miaadrajabi.menuxmlparser.data.source.AssetsXmlDataSource
import com.miaadrajabi.menuxmlparser.data.source.LocalXmlDataSource
import com.miaadrajabi.menuxmlparser.data.xml.DynamicMenusXml
import com.miaadrajabi.menuxmlparser.domain.model.DynamicMenus
import com.miaadrajabi.menuxmlparser.domain.model.MenuGroup

class MenuRepository(
    private val assets: AssetsXmlDataSource,
    private val local: LocalXmlDataSource
) {
    fun load(): DynamicMenus {
        val fromLocal = local.read()
        val xml: DynamicMenusXml = fromLocal ?: assets.read()
        return xml.toDomain()
    }

    fun save(domain: DynamicMenus) {
        // Convert domain back to XML DTO shape
        val xml = domain.toXml()
        local.write(xml)
    }
}

// Reverse mappers for persistence
private fun DynamicMenus.toXml(): DynamicMenusXml = DynamicMenusXml(
    menuGroups = groups.map { it.toXml() }
)

private fun MenuGroup.toXml(): com.miaadrajabi.menuxmlparser.data.xml.MenuGroupXml {
    val topLevel = topLevelMenus.map { it.toXml() }
    return com.miaadrajabi.menuxmlparser.data.xml.MenuGroupXml(
        clazz = null,
        enTitle = englishTitle,
        enable = isEnabled.toString(),
        faTitle = persianTitle,
        id = id,
        topLevelMenus = topLevel
    )
}

private fun com.miaadrajabi.menuxmlparser.domain.model.TopLevelMenu.toXml(): com.miaadrajabi.menuxmlparser.data.xml.TopLevelMenuXml =
    com.miaadrajabi.menuxmlparser.data.xml.TopLevelMenuXml(
        enTitle = englishTitle,
        enable = isEnabled.toString(),
        expandable = isExpandable?.toString(),
        faTitle = persianTitle,
        icon = icon,
        id = id,
        services = services.map { it.toXml() },
        items = items.map { it.toXml() }
    )

private fun com.miaadrajabi.menuxmlparser.domain.model.MenuItem.toXml(): com.miaadrajabi.menuxmlparser.data.xml.MenuItemXml =
    com.miaadrajabi.menuxmlparser.data.xml.MenuItemXml(
        clazz = clazz,
        enTitle = englishTitle,
        enable = isEnabled.toString(),
        faTitle = persianTitle,
        id = id,
        ussdGetSimCharge = ussdGetSimCharge,
        ussdGetSimNum = ussdGetSimNum,
        usssdChargeCommand = usssdChargeCommand,
        children = children.map { it.toXml() },
        services = services.map { it.toXml() }
    )

private fun com.miaadrajabi.menuxmlparser.domain.model.ServiceDescriptor.toXml(): com.miaadrajabi.menuxmlparser.data.xml.ServiceDescriptorXml =
    com.miaadrajabi.menuxmlparser.data.xml.ServiceDescriptorXml(
        id = id,
        service = service,
        serviceAmount = serviceAmount,
        categoryId = categoryId,
        hasCount = hasCount?.toString(),
        load = load,
        providerId = providerId,
        support = support,
        algorithmType = algorithmType,
        companyNameEn = companyNameEn,
        companyNameFa = companyNameFa
    )


